### PR TITLE
Implementing `asfloat` using `bit_cast`

### LINKED ISSUE #70098: Implement the `asfloat` HLSL Function
- [ ] Link `asfloat` clang builtin (__builtin_bit_cast) with `hlsl_intrinsics.h``
- [ ] Add codegen tests to `clang/test/CodeGenHLSL/builtins/asfloat.hlsl`
- [ ] Add sema tests to `clang/test/SemaHLSL/BuiltIns/asfloat-errors.hlsl`

## DirectX

There were no DXIL opcodes found for `asfloat`.

### Example 1
```hlsl
//dxc asfloat_test.hlsl -T lib_6_8  -enable-16bit-types -spirv -fspv-target-env=universal1.5 -fcgl -O0

export float4 fn(float p1) {
    return asfloat(p1);
}
```
 ### Example 2
```hlsl
//dxc asfloat_1_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export float4 fn(uint p1) {
    return asfloat(p1);
}
```
 ### Example 3
```hlsl
//dxc asfloat_2_test.hlsl -T lib_6_8 -enable-16bit-types -O0

export float4 fn(int p1) {
    return asfloat(p1);
}
```
## HLSL:

Interprets the bit pattern of *x* as a floating-point number.

| ret asfloat(*x*) |
|------------------|

 

## Parameters

| Item                                                   | Description                        |
|--------------------------------------------------------|------------------------------------|
| <span id="x"></span><span id="X"></span>*x*<br/> | \[in\] The input value.<br/> |

 

## Return Value

The input interpreted as a floating-point number.

## Type Description

| Name  | [**Template Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                                                  | [**Component Type**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)                                                         | Size                           |
|-------|----------------------------------------------------------------------------------------------------------------|------------------------------------------------------------------------------------------------------------------------|--------------------------------|
| *x*   | [**scalar**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md), **vector**, or **matrix** | [**float**](/windows/desktop/WinProg/windows-data-types), [**int**](/windows/desktop/WinProg/windows-data-types), [**uint**](/windows/desktop/WinProg/windows-data-types) | any                            |
| *ret* | same as input *x*                                                                                              | [**float**](/windows/desktop/WinProg/windows-data-types)                                                                                | same dimension(s) as input *x* |

 

## Function Overloads

<dl> `float&lt;x&gt; asfloat(float&lt;x&gt; value);`  
`float&lt;x&gt; asfloat(int&lt;x&gt; value);`  
`float&lt;x&gt; asfloat(uint&lt;x&gt; value);`  
</dl>

## Minimum Shader Model

This function is supported in the following shader models.

| Shader Model                                                        | Supported |
|---------------------------------------------------------------------|-----------|
| [Shader Model 4](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm4.md) and higher shader models | yes       |
| [Shader Model 3 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm3.md)           | no        |
| [Shader Model 2 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm2.md)           | no        |
| [Shader Model 1 (DirectX HLSL)](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-sm1.md)           | no        |

 

## Remarks

Older compilers incorrectly allowed `asfloat(bool)`, but note that bool inputs are not supported.

## See also

<dl> <dt>

[**Intrinsic Functions (DirectX HLSL)**](https://github.com/MicrosoftDocs/win32/blob/docs/desktop-src//direct3dhlsl/dx-graphics-hlsl-intrinsic-functions.md)
</dt> </dl>

### PR DESCRIPTION
**Pull request by joaosaffran**
This PR is implementing `asfloat` for HLSL.

Fixes: #70098

### CODE REVIEW

File: clang/lib/Headers/hlsl/hlsl_intrinsics.h [hunk 1]
```diff
@@ -361,6 +361,23 @@ bool any(double3);
 _HLSL_BUILTIN_ALIAS(__builtin_hlsl_any)
 bool any(double4);
 
+//===----------------------------------------------------------------------===//
+// asfloat builtins
+//===----------------------------------------------------------------------===//
+
+/// \fn uint asfloat(T Val)
+/// \brief Interprets the bit pattern of x as an unsigned integer.
```
Comment (llvm-beanz) [inline clang/lib/Headers/hlsl/hlsl_intrinsics.h:369]: This looks copy-pasted and not updated.

Comment (joaosaffran) [reply clang/lib/Headers/hlsl/hlsl_intrinsics.h]: Missed that, thanks for pointing out

**Review by python3kgae — APPROVED** [review]

File: clang/test/SemaHLSL/BuiltIns/asfloat-errors.hlsl [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+}
+
+
+float test_float_half(bool p1) {
+    return asfloat(p1);
+    // expected-error@hlsl/hlsl_intrinsics.h:* {{no matching function for call to 'bit_cast'}}
+    // expected-note@-2 {{in instantiation of function template specialization 'hlsl::asfloat<bool>'}}
+    // expected-note@hlsl/hlsl_detail.h:* {{candidate template ignored: could not match 'vector<bool, N>' against 'bool'}}
+    // expected-note@hlsl/hlsl_detail.h:* {{candidate template ignored: substitution failure [with U = float, T = bool]: no type named 'Type'}}
+}
```
Comment (farzonl) [inline clang/test/SemaHLSL/BuiltIns/asfloat-errors.hlsl:35]: add a new line

**Review by farzonl — APPROVED** [review]

**Comment by joaosaffran** [comment]
Can someone merge this, please :)

**Pull request merged**